dmoj/problem.py

```diff
@@ -45,6 +51,90 @@ def __init__(self, problem_id, time_limit, memory_limit):
             raise InvalidInitException(str(e))
 
         self.problem_data.archive = self._resolve_archive_files()
+        self._resolve_test_cases()
+
+    def _match_test_cases(self, filenames, input_case_pattern, output_case_pattern, case_points):
+        def try_match_int(match, group):
+            try:
+                val = match.group(group)
+            except IndexError:
+                return None
+
+            try:
+                return int(val)
+            except (ValueError, TypeError):
+                return val
+
+        def parse_position(pattern, filename):
+            match = pattern.match(filename)
+            if not match:
+                return None
+
+            # Allow batches and case numbers to be alphanumeric, in which case we will sort them lexicographically.
+            # Still attempt to process them as integers first, though, since most problems will use this format.
+            return try_match_int(match, 'batch'), try_match_int(match, 'case')
+
+        # Match all cases with the same (batch, position) mapping.
+        groups = defaultdict(list)
+        batch_ids = set()
+        for a in filenames:
+            a_parse = parse_position(input_case_pattern, a)
+            for b in filenames:
+                b_parse = parse_position(output_case_pattern, b)
+                if a_parse == b_parse and a_parse is not None:
+                    batch, case = a_parse
+                    assert case is not None
+                    if batch is not None:
+                        batch_ids.add(batch)
+                    groups[batch or case].append((case, a, b))
+
+        test_cases = []
+        for batch_or_case_id in sorted(groups.keys()):
+            if batch_or_case_id in batch_ids:
+                test_cases.append({
+                    'batched': [{
+                        'in': input_file,
+                        'out': output_file,
+                    } for _, input_file, output_file in sorted(groups[batch_or_case_id])],
+                    'points': next(case_points),
+                })
+            else:
+                assert len(groups[batch_or_case_id]) == 1
```

Comment:
Maybe we should raise real exceptions. Imagine running the judge with `python -O`.